contracts/money-market/PriceOracle/PriceOracleProxyUSD.sol

```diff
@@ -0,0 +1,184 @@
+    }
+
+    /*** Internal fucntions ***/
+
+    /**
+     * @notice Get price from ChainLink
+     * @param aggregator The ChainLink aggregator to get the price of
+     * @return The price
+     */
+    function getPriceFromChainlink(AggregatorV3Interface aggregator)
+        internal
+        view
+        returns (uint256)
+    {
+        (, int256 price, , , ) = aggregator.latestRoundData();
+        require(price > 0, "invalid price");
+
+        // Extend the decimals to 1e18.
+        return mul_(uint256(price), 10**(18 - uint256(aggregator.decimals())));
+    }
+
+    /**
+     * @notice Get price from v1 price oracle
+     * @param bTokenAddress The BToken address
+     * @return The price
+     */
+    function getPriceFromV1(address bTokenAddress)
+        internal
+        view
+        returns (uint256)
+    {
+        address underlying = BErc20(bTokenAddress).underlying();
+        return v1PriceOracle.assetPrices(underlying);
+    }
+
+    /*** Admin or guardian functions ***/
+
+    event AggregatorUpdated(
+        address bTokenAddress,
+        address source,
+        AggregatorBase base
+    );
+    event SetGuardian(address guardian);
+    event SetAdmin(address admin);
+
+    /**
+     * @notice Set guardian for price oracle proxy
+     * @param _guardian The new guardian
+     */
+    function _setGuardian(address _guardian) external {
+        require(msg.sender == admin, "only the admin may set new guardian");
+        guardian = _guardian;
+        emit SetGuardian(guardian);
+    }
+
+    /**
+     * @notice Set admin for price oracle proxy
+     * @param _admin The new admin
+     */
+    function _setAdmin(address _admin) external {
+        require(msg.sender == admin, "only the admin may set new admin");
+        admin = _admin;
+        emit SetAdmin(admin);
+    }
+
+    /**
+     * @notice Set ChainLink aggregators for multiple bTokens
+     * @param bTokenAddresses The list of bTokens
+     * @param sources The list of ChainLink aggregator sources
+     * @param bases The list of ChainLink aggregator bases
+     */
+    function _setAggregators(
+        address[] calldata bTokenAddresses,
+        address[] calldata sources,
+        AggregatorBase[] calldata bases
+    ) external {
+        require(
+            msg.sender == admin || msg.sender == guardian,
+            "only the admin or guardian may set the aggregators"
+        );
+        require(
+            bTokenAddresses.length == sources.length &&
+                bTokenAddresses.length == bases.length,
+            "mismatched data"
+        );
+        for (uint256 i = 0; i < bTokenAddresses.length; i++) {
+            if (sources[i] != address(0)) {
+                require(
+                    msg.sender == admin,
+                    "guardian may only clear the aggregator"
+                );
+            }
+            aggregators[bTokenAddresses[i]] = AggregatorInfo({
+                source: AggregatorV3Interface(sources[i]),
```

Comment:
How would you suggest we do this? My first thought would be to get the price of the asset immediately after and validate that the price is greater than 0. Does that align with you ideas or do you see another route being better? 
@Gornutz Also curious on your thoughts about input validation as it would add extra costs for setting.

Reply:
@0xSpraggins I thought checking the price > 0 could solve that as well. As for the extra costs, I assume it should not be much of a worry since it's the admin that does the update and not a user-facing function. But it will be nice to hear what @Gornutz thinks.